modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1700,19 +1700,33 @@ void GDScriptAnalyzer::resolve_for(GDScriptParser::ForNode *p_for) {
 								reduced = Vector3i(args[0], args[1], args[2]);
 								break;
 						}
+
 						p_for->list->is_constant = true;
 						p_for->list->reduced_value = reduced;
-					}
-				}
+						p_for->list->set_datatype(type_from_variant(reduced, p_for->list));
+					} else {
+						Variant::Type range_replacement = Variant::NIL;
+						switch (call->arguments.size()) {
+							case 1:
+								range_replacement = Variant::INT;
+								break;
+							case 2:
+								range_replacement = Variant::VECTOR2I;
+								break;
+							case 3:
+								range_replacement = Variant::VECTOR3I;
+								break;
+						}
 
-				if (p_for->list->is_constant) {
-					p_for->list->set_datatype(type_from_variant(p_for->list->reduced_value, p_for->list));
-				} else {
-					GDScriptParser::DataType list_type;
-					list_type.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
-					list_type.kind = GDScriptParser::DataType::BUILTIN;
-					list_type.builtin_type = Variant::ARRAY;
-					p_for->list->set_datatype(list_type);
+						call->function_name = Variant::get_type_name(range_replacement);
+						callee->name = Variant::get_type_name(range_replacement);
```

Comment:
I'm not really fond of changing what's in the parser tree like this. Mostly because if you try to print the tree afterwards (which I do for debugging) you'll get a call to `Vector2i` instead of `range` which might get confusing, even if you know about this.

Reply:
The proper solution would be to flag this call so the compiler can emit the instruction for constructor instead of a call to `range`.